Feature/routingresult failurereason

dropdevcoding opened the pull request:
Made Routes be kept in RoutingResult instead of plain targets
Added failure reason property to RoutingResult
Removed default target because it should be a more domain specific thing how errors should be handled if no defined route matches

dropdevcoding committed:
changed routingresult to return route instead of target only
introduced routingresult failuretypes
changed Psr7Router to create failure RoutingResults with appropriate failure code and the topmost route which could have matched by either using another method or different params

dropdevcoding committed:
added unit tests for routing result to ensure that route is returned correctly if set for failure

shochdoerfer commented on src/bitExpert/Pathfinder/Psr7Router.php:
```diff
@@ -35,52 +35,61 @@ public function match(ServerRequestInterface $request)
                 // remove all elements which should not be set in the request,
                 // e.g. the matching url string as well as all numeric items
                 $params = $this->mapParams($urlVars);
                 $identifier = $this->getRouteIdentifier($route);
```
Call not needed because $identifier is already set?

dropdevcoding committed:
removed duplicate call

dropdevcoding committed:
minor doc block corrections
changed configurationexception to invalidargumentexceptions
corrected unittests in favor of new exception types

dropdevcoding committed:
composer update
